### PR TITLE
[processing] port some parameter definition widgets to C++

### PR DESCRIPTION
**Pull request by alexbruy**
## Description

Port CRS, point and expression parameter definition widgets to C++. This allows to use native API to create parameter definition dialogs in modeler.

### CODE REVIEW

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 1]
```diff
@@ -1533,11 +1564,69 @@ QgsProcessingAbstractParameterDefinitionWidget *QgsProcessingFileWidgetWrapper::
...
+      }
+      else if ( const QgsProcessingParameterVectorLayer *definition = dynamic_cast< const QgsProcessingParameterVectorLayer * >( widgetContext.model()->parameterDefinition( it.value().parameterName() ) ) )
+      {
+        mParentLayerComboBox-> addItem( definition->description(), definition->name() );
+        if ( !initialParent.isEmpty() && initialParent == definition->name() )
+        {
+          mParentLayerComboBox->setCurrentIndex( mParentLayerComboBox->count() - 1 );
+        }
+      }
+    }
```
Comment (nyalldawson) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: The other definition widgets with parent links have this kind of logic here:
```
if ( mParentLayoutComboBox->count() == 0 && !initialParent.isEmpty() )
  {
    // if no parent candidates found, we just add the existing one as a placeholder
    mParentLayoutComboBox->addItem( initialParent, initialParent );
    mParentLayoutComboBox->setCurrentIndex( mParentLayoutComboBox->count() - 1 );
  }
```
I think we'll need the same here

**Review by nyalldawson — APPROVED** [review]
Great!

**Comment by nyalldawson** [comment]
Sorry - one last thing. All the ported definition widgets are covered by unit tests like https://github.com/qgis/QGIS/blob/master/tests/src/gui/testprocessinggui.cpp#L794 -- could you add them for these too?

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 2]
```diff
@@ -2527,11 +2621,37 @@ void QgsProcessingPointPanel::pointPicked()
...
+QgsProcessingPointParameterDefinitionWidget::QgsProcessingPointParameterDefinitionWidget( QgsProcessingContext &context, const QgsProcessingParameterWidgetContext &widgetContext, const QgsProcessingParameterDefinition *definition, const QgsProcessingAlgorithm *algorithm, QWidget *parent )
+  : QgsProcessingAbstractParameterDefinitionWidget( context, widgetContext, definition, algorithm, parent )
+{
+  QVBoxLayout *vlayout = new QVBoxLayout();
+  vlayout->setMargin( 0 );
+  vlayout->setContentsMargins( 0, 0, 0, 0 );
+
+  vlayout->addWidget( new QLabel( tr( "Default value" ) ) );
+
+  mDefaultLineEdit = new QLineEdit();
```
[the remark this answers is not in the document]
Comment (alexbruy) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Done

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 1 above)
Comment (alexbruy) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Added.

**Comment by alexbruy** [comment]
Tests added too.

**Review by nyalldawson — APPROVED** [review]
Looks great!

**Pull request merged**